src/mlpack/methods/bias_svd/bias_svd_function_impl.hpp

```diff
@@ -0,0 +1,411 @@
+/**
+ * @file bias_svd_function_impl.hpp
+ * @author Siddharth Agrawal
+ * @author Wenhao Huang
+ *
+ * An implementation of the BiasSVDFunction class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_BIAS_SVD_BIAS_FUNCTION_SVD_IMPL_HPP
+#define MLPACK_METHODS_BIAS_SVD_BIAS_FUNCTION_SVD_IMPL_HPP
+
+#include "bias_svd_function.hpp"
+#include <mlpack/core/math/make_alias.hpp>
+
+namespace mlpack {
+namespace svd {
+
+template <typename MatType>
+BiasSVDFunction<MatType>::BiasSVDFunction(const MatType& data,
+                                          const size_t rank,
+                                          const double lambda) :
+    data(math::MakeAlias(const_cast<MatType&>(data), false)),
+    rank(rank),
+    lambda(lambda)
+{
+  // Number of users and items in the data.
+  numUsers = max(data.row(0)) + 1;
+  numItems = max(data.row(1)) + 1;
+
+  // Initialize the parameters.
+  // The last row in initialPoint saves use/item bias.
+  initialPoint.randu(rank + 1, numUsers + numItems);
+}
+
+template<typename MatType>
+void BiasSVDFunction<MatType>::Shuffle()
+{
+  data = data.cols(arma::shuffle(arma::linspace<arma::uvec>(0, data.n_cols - 1,
+      data.n_cols)));
+}
+
+template <typename MatType>
+double BiasSVDFunction<MatType>::Evaluate(const arma::mat& parameters) const
+{
+  // The cost for the optimization is as follows:
+  //    f(u, v, p ,q) = sum((rating(i, j) - p(i) - q(j) - u(i).t() * v(j))^2)
+  // The sum is over all the ratings in the rating matrix.
+  // 'i' points to the user and 'j' points to the item being considered.
+  // The regularization term is added to the above cost, where the vectors u(i)
+  // and v(j), bias p(i) and q(j) are regularized for each rating they
+  // contribute to.
+
+  double cost = 0.0;
+
+  for (size_t i = 0; i < data.n_cols; i++)
+  {
+    // Indices for accessing the the correct parameter columns.
+    const size_t user = data(0, i);
+    const size_t item = data(1, i) + numUsers;
+
+    // Calculate the squared error in the prediction.
+    const double rating = data(2, i);
+    const double userBias = parameters(rank, user);
+    const double itemBias = parameters(rank, item);
+    double ratingError = rating - userBias - itemBias -
+        arma::dot(parameters.col(user).subvec(0, rank - 1),
+                  parameters.col(item).subvec(0, rank - 1));
```

Comment:
I forgot to say that you could avoid doing every time `.subvec(0, rank - 1)`. In order to do that you can introduce another matrix `arma::mat userItemMatrix(parameter.memptr(), rank, numUsers + numItems, false, false)`. I just wanted to make sure that you know.

Reply:
I'm afraid this doesn't work, because in armadillo matrix is stored in column-major ordering.

Reply:
Indeed, `armadillo` stores matrices in a column-major order, I didn't think of that. However you could reshape the `parameters` matrix to `1 x ((rank + 1) * (numUsers + numItems))`. However, not sure that it is worth to do.

Reply:
@lozhnikov I don't understand how reshaping `parameters` can solve this. Can you provide a code example?

Reply:
Ah I understand now. Thanks for the example. In my view both implementations are good, but the implementation you suggested is more clear. It could take some time to modify. I think for now we can stay with the current implementation, and write down some code comments as a reminder so we can improve this in another PR later. What do you think?

Reply:
Yeah, I think even the comment is unnecessary. Both implementations are good.